File: Editorial/25_Reverse_Nodes_in_k_Group/C/1_Recursion.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* reverseLinkedList(struct ListNode* head, int k) {
    struct ListNode* new_head = NULL;
    struct ListNode* ptr = head;
    while (k > 0) {
        struct ListNode* next_node = ptr->next;
        ptr->next = new_head;
        new_head = ptr;
        ptr = next_node;
        k--;
    }
    return new_head;
}
struct ListNode* reverseKGroup(struct ListNode* head, int k) {
    int count = 0;
    struct ListNode* ptr = head;
    while (count < k && ptr != NULL) {
        ptr = ptr->next;
        count++;
    }
    if (count == k) {
        struct ListNode* reversedHead = reverseLinkedList(head, k);
        head->next = reverseKGroup(ptr, k);
        return reversedHead;
    }
    return head;
}

File: Editorial/0000-0100/0025_Reverse_Nodes_in_k_Group/Go/1_Recursion.go
func reverseLinkedList(head *ListNode, k int) *ListNode {
    var newHead *ListNode = nil
    ptr := head
    for k > 0 {
        nextNode := ptr.Next
        ptr.Next = newHead
        newHead = ptr
        ptr = nextNode
        k--
    }
    return newHead
}

func reverseKGroup(head *ListNode, k int) *ListNode {
    count := 0
    ptr := head
    for count < k && ptr != nil {
        ptr = ptr.Next
        count++
    }
    if count == k {
        reversedHead := reverseLinkedList(head, k)
        head.Next = reverseKGroup(ptr, k)
        return reversedHead
    }
    return head
}
